src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerOptionsExtensions.cs

```diff
@@ -17,8 +15,10 @@ internal static class AnalyzerOptionsExtensions
 
         private const string DotnetAnalyzerDiagnosticSeverityKey = DotnetAnalyzerDiagnosticPrefix + "." + SeveritySuffix;
 
+        private static readonly ConcurrentDictionary<string, string> s_categoryToSeverityKeyMap = new ConcurrentDictionary<string, string>();
+
         private static string GetCategoryBasedDotnetAnalyzerDiagnosticSeverityKey(string category)
-            => $"{DotnetAnalyzerDiagnosticPrefix}.{CategoryPrefix}-{category}.{SeveritySuffix}";
+            => s_categoryToSeverityKeyMap.GetOrAdd(category, category => $"{DotnetAnalyzerDiagnosticPrefix}.{CategoryPrefix}-{category}.{SeveritySuffix}");
```

Comment:
>category [](http://example.com/codeflow?start=61&length=8)

Who controls these values? Is it some predefined set, or is it something that is coming from users?

Reply:
@AlekseyTs This method is all about constructing the key. The values for these keys are read from `.editorconfig`/`.globalconfig` from `AnalyzerOptions`

Reply:
So, the values are user controlled and a user will be able to grow ```s_categoryToSeverityKeyMap``` dictionary without any limit. Is this correct?

Reply:
It looks like the dictionary is going to be alive for as long as the process is alive. It could be shared between different compilations. Theoretically, it might be possible to create more and more compilations to keep the dictionary grow. If that is the case, I would be more comfortable to have an explicit cap on the amount of key cached (for example, keep certain amount of most recently used keys), or have the cache tied to a compilation (goes away once compilation goes away).

Reply:
Sharing between compilations was actually intended per review comment https://github.com/dotnet/roslyn/pull/72163#discussion_r1494758606

Different compilations will still share most categories. I wouldn't expect all categories out there to exceed 20 or something.

Reply:
> I wouldn't expect all categories out there to exceed 20 or something.

In a perfect world, yes. But we should also think about a worst case scenario. If we are certain about the amount of categories we will have to support, let's explicitly cap the size of the cache, let's say to 50 items. For example, we have ```ConcurrentLruCache``` type.